### PR TITLE
test: test all actions for out-of-gas results

### PR DESCRIPTION
**Pull request by jakmeier**
Any action can fail due to gas limit exceeded or gas attached exceeded.
We must preserve the exact result, including how much gas is burnt.

In #8044, we missed such a case and accidentally changed the burnt
gas value if it fails in the wrong place and only exceeds the attached
gas, not the gas burnt limit.

Tests added here cover this case for all existing action gas costs and
would catch this specific error in the future.

Not covered yet are ext costs and the case where we add new action
costs. We can bandage aid by adding similar tests for ext costs but we
also have to come up with a more general prevention.

### CODE REVIEW

**Comment by jakmeier** [comment]
ping @Longarithm 
would be nice to get these tests reviewed, they are more of the sort that you already reviewed for the accidental protocol change some time ago

File: runtime/near-vm-logic/src/tests/gas_counter.rs [hunk 1]
```diff
@@ -350,6 +354,332 @@ fn create_promise_dependency(logic: &mut VMLogic) -> Result<(), VMLogicError> {
     Ok(())
 }
 
+#[test]
+fn out_of_gas_new_data_receipt_byte() {
+    check_action_gas_exceeds_limit(ActionCosts::new_data_receipt_byte, 11, value_return);
```
Comment (Longarithm) [inline runtime/near-vm-logic/src/tests/gas_counter.rs:359]: I would add comments for `check_action_gas_exceeds_limit` and `check_action_gas_exceeds_attached` explaining the relations between `exercise_action` and other calls - e.g. "it is guaranteed that `exercise_action` triggers charging `cost` `num_action_paid` times".

Also we could have a comment in the beginning of the file explaining the general scheme of the tests:
* defining specific "exercise_action" which should trigger charging some specific cost
* call `check_action_gas_exceeds_limit`
* call `check_action_gas_exceeds_attached`

Comment (jakmeier) [reply runtime/near-vm-logic/src/tests/gas_counter.rs:403]: good point, I added a bunch of new comments in the latest commit, please take a look

File: runtime/near-vm-logic/src/tests/gas_counter.rs [hunk 2]
```diff
@@ -338,18 +389,359 @@ fn out_of_gas_new_data_receipt() {
...
+    check_action_gas_exceeds_attached(
+        ActionCosts::add_function_call_key_byte,
+        7,
+        expect!["236200046312 burnt 10000000000000 used"],
+        promise_function_key,
+    );
+}
+
+/// function to trigger base + 7 bytes action costs for adding a new function
+/// call access key to an accont (7 is arbitrary)
```
Comment (Longarithm) [inline runtime/near-vm-logic/src/tests/gas_counter.rs:702]:
```suggestion
/// call access key to an account (7 is arbitrary)
```

File: runtime/near-vm-logic/src/tests/gas_counter.rs [hunk 3]
```diff
@@ -338,18 +389,359 @@ fn out_of_gas_new_data_receipt() {
...
+    check_action_gas_exceeds_attached(
+        ActionCosts::function_call_byte,
+        40,
+        expect!["2444873079439 burnt 10000000000000 used"],
+        cross_contract_call_gas_weight,
+    );
+}
+
+/// function to trigger base + 40 bytes function call action costs (40 is 26 +
+/// 14  which are arbitrary)
```
Comment (Longarithm) [inline runtime/near-vm-logic/src/tests/gas_counter.rs:550]:
```suggestion
/// 14 which are arbitrary)
```

**Review by Longarithm — APPROVED** [review]
Thank you! Sorry I forgot to approve it just after our discussion.

**Comment by jakmeier** [comment]
huh, I completely forgot about this PR, I thought it was already merged :sweat_smile:

**Comment by jakmeier** [comment]
actually, let me wait for https://github.com/near/nearcore/pull/8375 to land, it will make some of the test nicer (and it will add merge conflicts)

**Pull request merged**